pkg/api/api.go

```diff
@@ -0,0 +1,304 @@
+	}
+}
+
+// // RegisterRing registers the ring UI page associated with the distributor for writes.
+func (a *API) RegisterRing(r *ring.Ring) {
+	a.registerRoute("/ring", r, false)
+}
+
+// RegisterStoreGateway registers the ring UI page associated with the store-gateway.
+func (a *API) RegisterStoreGateway(s *storegateway.StoreGateway) {
+	a.registerRoute("/store-gateway/ring", http.HandlerFunc(s.RingHandler), false)
+}
+
+// RegisterCompactor registers the ring UI page associated with the compactor.
+func (a *API) RegisterCompactor(c *compactor.Compactor) {
+	a.registerRoute("/compactor/ring", http.HandlerFunc(c.RingHandler), false)
+}
+
+// RegisterQuerier registers the Prometheus routes supported by the
+// Cortex querier service. Currently this can not be registered simultaneously
+// with the QueryFrontend.
+func (a *API) RegisterQuerier(queryable storage.Queryable, engine *promql.Engine, distributor *distributor.Distributor) {
+	api := v1.NewAPI(
+		engine,
+		queryable,
+		querier.DummyTargetRetriever{},
+		querier.DummyAlertmanagerRetriever{},
+		func() config.Config { return config.Config{} },
+		map[string]string{}, // TODO: include configuration flags
+		func(f http.HandlerFunc) http.HandlerFunc { return f },
+		func() v1.TSDBAdmin { return nil }, // Only needed for admin APIs.
+		false,                              // Disable admin APIs.
+		a.logger,
+		querier.DummyRulesRetriever{},
+		0, 0, 0, // Remote read samples and concurrency limit.
+		regexp.MustCompile(".*"),
+		func() (v1.RuntimeInfo, error) { return v1.RuntimeInfo{}, errors.New("not implemented") },
+		&v1.PrometheusVersion{},
+	)
+
+	promRouter := route.New().WithPrefix(a.cfg.ServerPrefix + a.cfg.PrometheusHTTPPrefix + "/api/v1")
+	api.Register(promRouter)
+	promHandler := fakeRemoteAddr(promRouter)
+
+	a.registerRoute(a.cfg.PrometheusHTTPPrefix+"/api/v1/read", querier.RemoteReadHandler(queryable), true, "GET")
+	a.registerRoute(a.cfg.PrometheusHTTPPrefix+"/api/v1/query", promHandler, true, "GET", "POST")
+	a.registerRoute(a.cfg.PrometheusHTTPPrefix+"/api/v1/query_range", promHandler, true, "GET", "POST")
+	a.registerRoute(a.cfg.PrometheusHTTPPrefix+"/api/v1/labels", promHandler, true, "GET", "POST")
+	a.registerRoute(a.cfg.PrometheusHTTPPrefix+"/api/v1/label/{name}/values", promHandler, true, "GET")
+	a.registerRoute(a.cfg.PrometheusHTTPPrefix+"/api/v1/series", promHandler, true, "GET", "POST", "DELETE")
+	a.registerRoute(a.cfg.PrometheusHTTPPrefix+"/api/v1/metadata", promHandler, true, "GET")
+
+	a.registerRoute("/api/v1/user_stats", http.HandlerFunc(distributor.UserStatsHandler), true)
+	a.registerRoute("/api/v1/chunks", querier.ChunksHandler(queryable), true)
+
+	// Legacy Routes
+	a.registerRoute("/user_stats", http.HandlerFunc(distributor.UserStatsHandler), true)
+	a.registerRoute("/chunks", querier.ChunksHandler(queryable), true)
+
+	legacyPromRouter := route.New().WithPrefix(a.cfg.ServerPrefix + a.cfg.LegacyHTTPPrefix + "/api/v1")
```

Comment:
Why do we need `ServerPrefix`? Before we were not adding it. Also the bext block of `registerRoute()` calls wasn't done before because already covered by `api.Register()` without `ServerPrefix`. Am I missing anything?

Reply:
I wrote it up in an issue here: #2411 

TLDR: the router used here handles the path separately from the weaveworks server router. If a prefix is set in the weaveworks server router, the promrouter will always return a 404 because it is not expecting the prefix. Anything directly registered with the weaveworks server router will not have this problem.